### PR TITLE
API 3922 Questionnaire Response Subject Search

### PR DESCRIPTION
**Pull request by that-guy-is-it**

### CODE REVIEW

File: patient-generated-data/src/main/resources/db/changelog/db.changelog-master.yaml [hunk 1]
```diff
@@ -141,6 +141,11 @@ databaseChangeLog:
                   type: datetime2
                   constraints:
                     - nullable: true
+              - column:
+                  name: subject
+                  type: varchar
+                  constraints:
+                    - nullable: true
```
Comment (apeirats-va) [inline patient-generated-data/src/main/resources/db/changelog/db.changelog-master.yaml:148]: You'll want to add a new changeset at the bottom with the new column and the correct data type. Don't modify previous changesets already merged to master.

Remember this is a changelog. We don't rebuild the DB from scratch on live environments, we just modify the existing database (hence the changelog). It's like adding new commits to git- we don't modify previous history, we add new changes to the history.

Liquibase keeps track of which changesets it has applied to the current DB. When it runs the changelog, if it runs into a changeset it knows it already ran, but it has been modified, it panics and bails.

File: patient-generated-data/src/main/java/gov/va/api/health/patientgenerateddata/questionnaireresponse/QuestionnaireResponseController.java [hunk 2]
```diff
@@ -56,11 +56,13 @@
...
                 .value("author", "author")
                 .dateAsInstant("authored", "authored")
+                .value("subject", "subject")
                 .get())
         .defaultQuery(returnNothing())
-        .rule(atLeastOneParameterOf("_id", "author", "authored"))
+        .rule(atLeastOneParameterOf("_id", "author", "authored", "subject"))
         .rule(parametersNeverSpecifiedTogether("_id", "author"))
         .rule(parametersNeverSpecifiedTogether("_id", "authored"))
+        .rule(parametersNeverSpecifiedTogether("_id", "subject"))
```
Comment (apeirats-va) [inline patient-generated-data/src/main/java/gov/va/api/health/patientgenerateddata/questionnaireresponse/QuestionnaireResponseController.java:65]: Let's replace these with:
`.rule(ifParameter("_id").thenForbidParameters("author", "authored", "subject"))`

File: patient-generated-data/src/main/resources/db/changelog/db.changelog-master.yaml [hunk 3]
```diff
@@ -219,3 +219,17 @@ databaseChangeLog:
...
+      id: 9
+      author: pteracuda
+      changes:
+        - addColumn:
+            tableName: QuestionnaireResponse
+            schemaName: app
+            columns:
+              - column:
+                  name: subject
+                  type: varchar
```
Comment (jblefkove-va) [inline patient-generated-data/src/main/resources/db/changelog/db.changelog-master.yaml:233]: `varchar` needs an explicit size, or the default (in SQL Server) will be too small for our purposes. The `local-db.sh` script fails to run.

Note that in the previous changeset `author` was changed from `varchar` to `varchar(255)`

**Comment by jblefkove-va** [comment]
`curl -s http://localhost:8095/r4/QuestionnaireResponse?subject=1017283180V801730`

LGTM

**Pull request merged**